ARM Support

joalen opened the pull request:
Like the issue I attached, I was able to add ARM support to the disassembly view in Jitwatch and can confirm with a bit of testing that it does fill out the tooltips with each of those ARM instructions. I also did the liberty of adding some unit tests to the ARM version of this so that it can work and verifies correctness.

Thanks again for your review!
~ Alen

joalen committed:
Update AssemblyReference.java

- Started implementing the SAXParser as this might be faster in parsing the XMLs for the assembly references.
- // TODO: Need to then let this become implemented within the static block after the else. Will probably need to compare the performance to see how effective SAXParser is to the original solution
- // TODO: Need to finish completing the Aarch64 XML reference files (tediously takes some time since ARM likes to give separate XMLs as opposed to a merged version)

joalen committed:
Update AssemblyReference.java

- added the SAXParser and removed the old logic to make way for the XML parsing for the assembly reference XMLs. Again, I'll probably need to measure the performance and see how effective this is.
- //TODO: Need to still complete compilation of Aarch64 instructions into a final XML (done manually with some automation)

joalen committed:
Exception Handling for XML parser and some compilation errors fixed

- forgot to add exception handling for the new XML parser
- needed to add in some imports from the org.xml section
- forgot to add the getter method for the mnemonic map

joalen committed:
Add aarch64reference XML, fixed AssemblyParserARM, and AssemblyReference

- I've gotten the aarch64reference XML added and it seems to be getting those assembly instructions, although a few things need to be sorted: ARM has the same instruction but for various purposes like for immediate or for register operations --> need to build out logic for that
- sorted the isRegister TODO found in the original ARM parser and adjusted based on the Aarch64 and Aarch32 standards found from ARM
- //TODO: need to make smarter parsing strategy for parsing duplicate instructions that refer to different meanings
- //TODO: refactor the ViewerAssembly code to not rely solely on X86 but also Aarch64 and other ISAs (if needed)

joalen committed:
Changed private to protected visibility for ASM_REF_PATH & NPEs found from specifying JIT Log through CLI

- found an unexpected NullPointerException when specifying the java -Djitwatch.logfile=testLog.log -jar ui/target/jitwatch-ui-shaded.jar if not specifying the sourceLocations or classLocations.
 - Change the reference path if ARM_64 is detected
- Turns out, past JDK 8, we don't get the liberty of "Disassembling #309 (a closed issue, but still not resolved). As a result, it always defaults to X86 parser through some trial-and-error in getting the log error printed
  - Will need to add a new parsing strategy to use the <info></info> tags to extract the architecture version from that (might be brittle but not sure if there aren't other ways of going about this)
    - //TODO: add a way to extract architecture information from the <info> tags of the header in the hotspot file

joalen committed:
Added strategy for JDK 11+ for finding proper architecture + create ARMRegisterEnum

- added a check to now parse the existing <info> to include the architecture from the <info> string for at least JDKs >11
- fixed extractRegisterName to be more specific to the ARM32 and ARM64 standards
- (for now temporary) -- added more clarification to the type of register that the ARM64 code may use; will need to adjust the decodeRegister method to move X86 and ARM instructions to their respective enums
- // TODO: continue to repair visualization for register operands getting populated in the tooltip
- // TODO: troubleshoot why ARM tooltip for certain instructions is not loading and resolve duplicate ARM instructions based on the instruction context

joalen committed:
Retransform AssemblyReference to fit more than one parsers + Moved X86 Registers to an Enum

- since we can support two or more AssemblyReferences in the ViewerAssembly, it's best to move the X86 register information into an Enum that can be accessible
- Now, the tooltip automatically obtains the reference for the Architecture of choice and resolves the todo for the IParser
- Lazy loading implementation allows for "dynamically" loading the AssemblyReference at will without doing it in static initialization

joalen committed:
Fixed inability to display the other operands + directives caused mangling in final AssemblyView

- only the first operand was successful in being viewed in the tooltip; however, needed to account for whitespaces after initial parsing of the first register
- added the Directives Enums to help resolve issues pertaining to directives mangled in the assembly
 - as a result, the AbstractAssemblyParser needed some conditions to check for these directives and include them; however, without, information for now.
- Needed to make sure that we check for the proper parser type and allow for picking up on the appropriate Mnemonic reference

joalen committed:
Finalized AARCH64 Aggregate XML Reference File

- Finalized and edited parenthetical versions of the AARCH64 instructions, where, in the same level, an <insFormat> tag acts as a regex pattern to help recognize the AARCH64 duplicated instructions.
- // TODO: add more to the parser to accept the insFormat for whenever we have a duplicate instruction -- most likely a HashMap or similar Dictionary type will come in use for this

joalen committed:
Added Parsing for duplicated AARCH64 instructions and removed unnecessary XML fields

- since AARCH64 has duplicate instructions for different purposes, I had to implement another Map that contains a MnemonicInfo object to contain the <brief> and the list of patterns (although now that I think about it, this poses a problem if EACH of the duplicate assembly instruction and its pattern have their own unique brief...will need to look into that)
- the aarch64reference.xml file was super large with 1M+ lines, so to reduce that, there were unnecessary fields (side effect of the PDF production it seems) that aren't needed for Jitwatch (some fields are kept if future features want to utilize some of those properties)
- preliminary testing for now, but hopefully a working solution will come with stronger AARCH64 assembly parsing

joalen committed:
Update the MnemonicEntry class for handling only mnemonic and brief + Fix regex patterns in aarch64reference.xml

- Certain reference XML patterns were not properly handled with the curly brace, causing errors to throw (was not as severe but instructions could be missed)
- MnemonicEntry class to make this the key and the actual mnemonic as the value due to the duplicate nature of AARCH64 instruction sets
- ViewerAssembly will parse the whole instruction line to obtain the pattern match if not found through the regular lookup map

joalen committed:
Broken regexes did not include for registers other than sp for ARM + improper lazy-loading caused no assemblies to load + better handling of instruction string

- there was a pending issue where not all ARM64 instructions decoded successfully; however, this is now resolved where it was first terrible handling of the lazy-loading (needed to do null rather than create an empty map) + only add clear mnemonics after insFormat or mnem tags
- certain ARM instructions did not match, which left them undefined. Now fixed by accounting for these extra registers
- when passing in the instruction strings, comments and hex addresses kept these strings from matching against the Patterns. Now, these are removed

joalen committed:
Add support for ARM directives + patching other ARM assembly line failures

- ARM directives added within the tooltip
- diagnosing a few bit of missed ARM assembly instructions missed and patching those patterns

joalen committed:
Add a few integration tests for ARM

- made some test methods to see how effective this ARM parser is; not all has been done, but will be done as soon as time permits

joalen committed:
Patched unit tests and regression tests for ARM parsing

- needed to make one block of an example Hotspot log consistent and ensures correctness; therefore, it should work
- created two additional tests to check for operand identification and common register names
- included some cases for ARM32 and ARM64
 - I'll make some tests for ARM 32 (which would be the assembly parsing ones) to add onto the ARM tests (right now, I just have ARM 64)

joalen committed:
Add Shift and Extend Tooltip

- turns out that there was missing information in the tooltip for when we encounter registers where we in-line shift/extend types. Now, this is included with these additions

joalen committed:
Removed old testOperand method

- found it clunky to try to make a test to try out shifts and extends. Tested it out on Jitwatch and it works for having the shifts and extends that were in-lined in an ARM instruction!!

chriswhocodes commented:
Hi @joalen when I run JITWatch with your patch on my Linux x86-64 system I get a ClassCastException.

I don't think the cast is needed on ViewerAssembly.java:253 ? Perhaps the abstract superclass can offer a Getter for the architecture?

Kind regards,

Chris

`Exception in thread "JavaFX Application Thread" java.lang.ClassCastException: class org.adoptopenjdk.jitwatch.model.assembly.x86.AssemblyParserX86 cannot be cast to class org.adoptopenjdk.jitwatch.model.assembly.arm.AssemblyParserARM (org.adoptopenjdk.jitwatch.model.assembly.x86.AssemblyParserX86 and org.adoptopenjdk.jitwatch.model.assembly.arm.AssemblyParserARM are in unnamed module of loader 'app')
	at org.adoptopenjdk.jitwatch.ui.triview.assembly.ViewerAssembly.decodeRegister(ViewerAssembly.java:253)
	at org.adoptopenjdk.jitwatch.ui.triview.assembly.ViewerAssembly.decodeOperand(ViewerAssembly.java:193)
	at org.adoptopenjdk.jitwatch.ui.triview.assembly.ViewerAssembly.getToolTip(ViewerAssembly.java:178)
	at org.adoptopenjdk.jitwatch.ui.triview.assembly.ViewerAssembly.setAssemblyMethod(ViewerAssembly.java:110)
	at org.adoptopenjdk.jitwatch.ui.triview.TriView.updateBytecodeAndAssembly(TriView.java:776)
	at org.adoptopenjdk.jitwatch.ui.triview.TriView.access$1100(TriView.java:76)
	at org.adoptopenjdk.jitwatch.ui.triview.TriView$14.run(TriView.java:1275)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:457)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:456)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:316)
	at java.base/java.lang.Thread.run(Thread.java:1583)
`

joalen committed:
ClassCastException due to unconditional casting on architecture

- forgot that there's a method to allow for retrieving the architecture without needing to do unconditional casting. This should allow Jitwatch to run and switch the context in between ARM or X86 systems.

joalen commented:
Hey @chriswhocodes 

Yup. I think you had a method for getting the architecture and I forgot that existed. I removed the typecasting in that conditional and it should just call that getter method instead. Hopefully that fixes the bug for X86 systems. Thanks again for catching it! 

Thank you!
~ Alen

joalen committed:
NPEs caught + begin mnemonic ARM fix + test case to find unparsed instructions

- I'll use the test case made to help check to see if any ARM instructions were left out and will have their information loaded in the tooltip
- Found some NPEs while running my test suite, which seems to come from certain instructions set to `null' which can cascade due to accessing null on a string
 - To help encounter and make debugging easier, I put it as errors as to help catch particular instances of these nulls if any
- it'll take me a lot of work but I can attempt to identify common groups of ARM instructions that fail at matching and fix their patterns (or implement new ones if necessary)

joalen commented:
Hi @chriswhocodes,

Yup, I was afraid of that! Certain instructions seem to fail in pattern matching and I think it's cause I forgot about the other variants of these ARM instruction formats. I fixed that issue, but wanted to do my due dilligence in checking other example files, where I made a test fixture to analyze all missing ARM instructions. I'll see if I can power through and get those mnemonics fixed and matched. 

Thanks again for your help. Very much appreciated!!
~ Alen

joalen committed:
Better Testing for ARM parsing with Jitwatch Examples + Hand-off to handler + more ARM mnemonic pattern fixes

- got a better testing method for ARM parsing in which we consume a file rather than take in a StringReader. This should help retain the same, real-world Hotspot log  as needed
- needed to unescape the double slashes since the XML parser will take this in as a character array and not just strings
- through testing with the new test harness, some ARM instructions fail to parse and now are fixed
- Within the AbstractAssemblyParser, rather than track for a particular line, let the AssemblyProcessor handle that line

joalen committed:
Remove Test Harness

- can be flaky due to relying on system compiler and can make maven tests run longer. This was really to investigate better ARM parsing and hopefully make compilation times faster

joalen committed:
JDK 12+ (and Valhalla) maximum entity element fix + pattern fixes for new ARM instructions

- currently doing tests on the Valhalla JVM and encountered a few bit of ARM instruction misses but not overwhelming in amount
- encountered a bug in which JDKs 12+ set an XML entity maximum, where the main reference file to ARM instructions exceeded that. These blocks were removed as a result + other ones that bloated the file.

joalen commented:
Ok @chriswhocodes 

I can confirm, that I did seem to fix the missing mnemonics from your example and others I caught as well when testing the Jitwatch examples for JDK 11 and the Valhalla JDK (made with base JDK 25 + binutils) on an M2 Macbook. See if everything is ok on your end!

Thanks again for your patience!
~ Alen

chriswhocodes commented:
Hi @joalen I've tested the latest patch and all the ARM instructions I examined had the correct tooltips :+1: 

Let me know if you think this patch is ready to merge or if you have further changes?

Many thanks,

Chris

chriswhocodes commented:
Many thanks for this contribution @joalen . I will add your name to the contributors and prepare a version 1.5 release.
